rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,734 @@
+    #[error("S3Error")]
+    S3Error,
+}
+
+#[derive(Clone)]
+struct Buckets(HashMap<&'static str, &'static str>);
+
+impl Buckets {
+    fn new() -> Self {
+        let mut names = HashMap::new();
+        names.insert("masks", "masks");
+        names.insert("masked-models", "masked-models");
+        names.insert("global-model", "global-model");
+        names.insert("global-masked-model", "global-masked-model");
+        Self(names)
+    }
+
+    fn masks(&self) -> &'static str {
+        self.0.get("masks").unwrap()
+    }
+
+    fn masked_models(&self) -> &'static str {
+        self.0.get("masked-models").unwrap()
+    }
+
+    fn global_model(&self) -> &'static str {
+        self.0.get("global-model").unwrap()
+    }
+
+    fn global_masked_model(&self) -> &'static str {
+        self.0.get("global-masked-model").unwrap()
+    }
+
+    fn names(&self) -> impl Iterator<Item = &&str> {
+        self.0.values()
+    }
+}
+
+#[derive(Clone)]
+pub struct S3Store {
+    s3_client: S3Client,
+    buckets: Buckets,
+}
+
+// API
+impl S3Store {
+    /// Create a new S3 store. The store creates and maintains 4 different buckets.
+    /// See [`Buckets`] for more information.
+    ///
+    /// The [`S3Store`] can also be used together with Minio via a custom region:
+    /// ```text
+    /// let region = Region::Custom {
+    ///    name: String::from("eu-east-3"),
+    ///    endpoint: String::from("http://127.0.0.1:9000"), // URL of minio
+    /// };
+    ///
+    /// let store = S3Store::new(region);
+    /// ```
+    pub fn new<P>(region: Region, credentials_provider: P) -> Self
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+    {
+        let dispatcher = HttpClient::new().expect("failed to create request dispatcher");
+        Self {
+            s3_client: S3Client::new_with(dispatcher, credentials_provider, region),
+            buckets: Buckets::new(),
+        }
+    }
+
+    /// Upload a [`Mask`].
+    pub async fn upload_mask(&self, key: &str, mask: &Mask) -> Result<(), StorageError> {
+        match self
+            .upload(self.buckets.masks(), key, mask.serialize())
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(_) => Err(StorageError::S3Error),
+        }
+    }
+
+    /// Upload a [`MaskedModel`].
+    pub async fn upload_masked_model(
+        &self,
+        key: &str,
+        masked_model: &MaskedModel,
+    ) -> Result<(), StorageError> {
+        // we don't need the keys of the masked models. we can retrieve all keys via `list objects`.
+        // maybe we need the number of objects in the bucket.
+        // the key is a random string, we could calculate the hash of a masked model and use this
+        // as a key but this would cost more time to calculate
+        match self
+            .upload(self.buckets.masked_models(), &key, masked_model.serialize())
+            .await
+        {
+            Ok(_) => Ok(()),
+            Err(_) => Err(StorageError::S3Error),
```

Comment:
It seems that we never use the returned value and that we always convert the error into a `StorageError`, so perhaps this could be done in the `upload` method itself? That way we could just do:

```rust
self.upload(self.buckets.masked_models(), &key, masked_model.serialize())
    .await?
```